lib/server/services/qualityControlFlag/QcFlagService.js

```diff
@@ -133,6 +150,90 @@ class QcFlagService {
+        if (!dataPassId && !simulationPassId) {
+            throw new BadParameterError('Cannot create QC flag without data pass or simulation pass');
+        } else if (dataPassId && simulationPassId) {
+            throw new BadParameterError('Cannot create QC flag for data pass and simulation pass simultaneously');
+        }
+
+        const queryBuilder = dataSource.createQueryBuilder();
+        if (dataPassId) {
+            queryBuilder.whereAssociation('dataPasses', 'id').is(dataPassId);
+        } else if (simulationPassId) {
+            queryBuilder.whereAssociation('simulationPasses', 'id').is(simulationPassId);
+        }
+
+        queryBuilder
+            .include({ association: 'effectivePeriods', attributes: [], required: true })
+            .include({ association: 'run', attributes: [] })
+            .include({ association: 'flagType', attributes: [] })
+            .include({ association: 'verifications', attributes: [] })
+            .set('attributes', (sequelize) => [
+                'runNumber',
+                'dplDetectorId',
+                [
+                    sequelize.literal('SUM(`effectivePeriods`.`to` - `effectivePeriods`.`from`) \
+                        / (COALESCE(run.time_trg_end, run.time_o2_end) - COALESCE(run.time_trg_start, run.time_o2_start)) '),
+                    'effectiveRunCoverage',
+                ],
+                [sequelize.col('`flagType`.bad'), 'bad'],
+                [
+                    sequelize.literal('COUNT( DISTINCT `QcFlag`.id ) - COUNT( DISTINCT `verifications`.flag_id )'),
+                    'missingVerificationsCount',
+                ],
+            ])
+            .groupBy('runNumber')
+            .groupBy('dplDetectorId')
+            .groupBy((sequlize) => sequlize.col('`flagType`.bad'));
+
+        const runDetectorSummaryList = (await QcFlagRepository.findAll(queryBuilder))
+            .map((summaryDb) =>
+                ({
+                    runNumber: summaryDb.runNumber,
+                    dplDetectorId: summaryDb.dplDetectorId,
+                    effectiveRunCoverage: parseFloat(summaryDb.get('effectiveRunCoverage'), 10),
+                    bad: Boolean(summaryDb.get('bad')),
+                    missingVerificationsCount: parseInt(summaryDb.get('missingVerificationsCount'), 10),
+                }));
+
+        const summary = {};
+        for (const runDetectorSunmmary of runDetectorSummaryList) {
+            const {
+                runNumber,
+                dplDetectorId,
+                bad,
+                effectiveRunCoverage,
+                missingVerificationsCount,
+            } = runDetectorSunmmary;
+
+            if (!summary[runNumber]) {
+                summary[runNumber] = {};
+            }
+
+            if (!summary[runNumber][dplDetectorId]) {
+                summary[runNumber][dplDetectorId] = {};
+            }
+
+            summary[runNumber][dplDetectorId]['missingVerificationsCount'] =
```

Comment:
No hardcoded strings please and it would be great when you extract these, to add a one-line definition

Reply:
done